Add Model Factories package

ObliviousHarmony opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This is an iteration on the work originally proposed in #26413. During development I realized that `@wordpress/api-fetch` would not be usable due to the missing `fetch` API in NodeJS. While there are polyfills available, I was concerned that the global nature of that package would cause issues if developers were to try to use the package in project that was already using the package. The other problem that I ran into was that the [fishery package](https://github.com/thoughtbot/fishery) had not yet reached 1.0.0 and lacked some of the support for asynchronous model creation we would need to easily generate the data on the API. That said, I was generally happy with the API proposed by my spike and felt it worthwhile to pursue once `fishery` reached 1.0.0.

This PR is a heavy one so I'd like to outline some of the changes here as a summary for help when reviewing:

- **Configuration Changes**: These are the notable changes I've made to package configuration.
    - I've added some ESLint plugins to the root `package.json` in order to lint TypeScript packages. I've also adjusted our build hooks so that they run on TS files.
    - I've added a root `tsconfig.json` to support building all of the TypeScript packages using `tsc —build`. This is mostly for the benefit of IDEs since we're using `lerna` for build orchestration. I've also aded a `tsconfig.base.json` to be extended by TypeScript packages and contain common configuration.
    - This PR adjusts our `lerna` configuration so that each package is able to handle its own build pipeline. This was done because this package uses TypeScript and so I wanted to be as explicit as possible when deciding how a given package should be compiled. The existing `tests/e2e/bin/build.js` file was adjusted accordingly and the `@woocommerce/e2e-environment`'s `package.json` was updated to use it.
- **Package Architecture**: These are the core components of the package.
    - **ModelRegistry**: Instead of having exported variables like proposed by the `fishery` package, I decided to hold all of the factories in a registry so that we can perform actions against them easily and avoid having any global state. The current application of this is being able to easily change the adapter each factory is using.
    - **Model**: At the heart of the package are the model definitions themselves. These classes should contain all of the configurable properties for the models on the server.
    - **ModelFactory**: This class extends the `fishery` factory in order to add a `create` and `createList` method that uses an adapter to create the data on the server.
    - **Adapters**: An adapter is an abstraction that allows us to swap out the way that factories generate their models. While our default is going to be creation via the API, my thinking here was that we might want to support creation via the database in cases where the quantity of data to generate is more significant and could negatively impact test performance.
        - **API Adapter**: The API adapter handles creating models using API requests. This adapter is very simple and does a POST to an endpoint with data returned from a transformation function that converts the Model class into the database request format.
             - **AxiosAPIService**: One of the things that mattered to me was that consumers be able to spend as little effort as possible configuring and consuming the package. To that end I chose to wrap the Axios client in an abstraction to keep setup as painfully simple as possible, as well as make writing unit tests easier.

Here is a small example of the package's configuration and usage:

```javascript
import {
    AdapterTypes,
    initializeUsingBasicAuth,
    ModelRegistry,
    registerSimpleProduct,
    SimpleProduct
} from '@woocommerce/model-factories';

// The ModelRegistry instance is where all of the factories and adapters are stored in an easy-to-access way.
const modelRegistry = new ModelRegistry()

// Call the register functions to add a kind of factory to the model registry.
// This will also add any adapters we've created for the factory, allowing it
// to be created on the server.
registerSimpleProduct( modelRegistry );

// Before you can use the included API adapter you need to initialize it using one of the utility methods.
// If you do not initialize the API adapters they will not be able to make requests to the API.
// Note that these utility functions only set up adapters that have been registered already
// and so further calls to `registeryXXX` functions will have adapters that aren't ready.
initializeUsingBasicAuth( modelRegistry, 'https://test.test/wp-json', 'admin', 'password' );
initializeUsingOAuth( modelRegistry, 'https://test.test/wp-json', 'consumer_key', 'consumer_secret' );

// In order to actually create the models on the server, each registered factory must have an adapter set.
// You can do this on a per-factory basis using
modelRegistry.changeFactoryAdapter( SimpleProduct, AdapterTypes.API );
// You can do this to all factories registered using
modelRegistry.changeAllFactoryAdapters( AdapterTypes.API );

// Once all of the initialization has been taken care of you can create models!
// Any fields that are not defined will be filled out by random data.
const product = await modelRegistry.getFactory( SimpleProduct ).create( { name: 'Test Product' } );
// You can now access the ID of the created model using `product.id`!

// You can also create models in bulk!
const poducts = await modelRegistry.getFactory( SimpleProduct ).createList( 5 );
// You now have an array of products to work with!
```

I've also got a few things I'd like feedback on as well:
- Should I move all of the tests files into a `tests` folder? I'm a fan of having them next to the source files but it might be more consistent to do what we're doing elsewhere.
- Are the `index.ts` files I've made sufficient for making it easy to import everything from the package on an as-needed basis?

_One noteworthy thing as well is that the `SimpleProduct` here does not have any fields other than the name and regular price. We can add the rest of the fields, as well as support associated fields, but I wanted to get something submitted for review._

### How to test the changes in this Pull Request:

1. Install all composer & npm dependencies
2. Run `npm run build:packages`
3. Run `npm run docker:up`
4. Run `npm run test:e2e`
5. Run `npm run docker:down`

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

ObliviousHarmony committed:
Initial configuration for e2e-factories package

ObliviousHarmony committed:
Excluded test files from the TypeScript build

ObliviousHarmony committed:
Added an interceptor to handle WooCommerce API authentication

ObliviousHarmony committed:
Adjusted the package to better support NPM distribution

ObliviousHarmony committed:
Removed build artifact

ObliviousHarmony committed:
Adjusted the APIAuthInterceptor to be more explicitly Axios-specific

ObliviousHarmony committed:
Fixed the usage of baseURL with the APIAuthInterceptor

ObliviousHarmony committed:
Added a service for interacting with the WordPress API

We've hid the use of Axios behind a service so that we're able to easily mock it out in factories as well as handle the creation and configuration of the client. This will make it easier on consumers in that they won't have to worry about things like authentication when using the API.

ObliviousHarmony committed:
Moved the Axios response/error transformations into an interceptor

ObliviousHarmony committed:
Added formatting requirements to the code style

ObliviousHarmony committed:
Adjusted the TypeScript configuration to use the --build flag

ObliviousHarmony committed:
Added an adapter to support creating models via the API

ObliviousHarmony committed:
Added a skeleton product factory for testing

ObliviousHarmony committed:
Changed the structure of error responses

In practice having the response class contained within the error feels a little backwards. We can instead have a structured APIError model that is contained in the APIResponse and have a consistent response format to consume.

ObliviousHarmony committed:
Fixed the URL being passed to the OAuth package

ObliviousHarmony committed:
Created a base ModelFactory for all factories to extend

ObliviousHarmony committed:
Simplified the package directory structure

ObliviousHarmony committed:
Added a registry to hold all of the factories and adapters

ObliviousHarmony committed:
Replaced the exported variables with functions to populate the registry

ObliviousHarmony committed:
Fixed the POST, PUT, and PATCH Axios service actions

ObliviousHarmony committed:
Added a utility for initializing all of the APIAdapters with a service

ObliviousHarmony committed:
Merge branch 'master' into package/e2e-factories/product

ObliviousHarmony committed:
Added missing @wordpress/eslint-plugin for factory package

ObliviousHarmony committed:
Adjusted the structure of the factory package for easier consumption

ObliviousHarmony committed:
Fixed a bug that would cause APIService failures to throw unhandled exceptions when creating

ObliviousHarmony committed:
Refactored interceptors to support different authentication styles

ObliviousHarmony committed:
Replaced the `createSimpleProduct` helper with a usage of the factory

ObliviousHarmony committed:
Made the model properties compliant with code style

ObliviousHarmony committed:
Adjusted the build:packages command to allow each package to build itself

ObliviousHarmony committed:
Moved the factory package init into a common utility

ObliviousHarmony committed:
Added support for faker to populate models with dummy data

ObliviousHarmony committed:
Removed unused adapter types

ObliviousHarmony committed:
Renamed the e2e-factories package to better reflect its usage-agnostic design

ObliviousHarmony committed:
Added a readme documenting the usage of the model factory package

ObliviousHarmony committed:
Added a registry for adapters and factories to be held in globally

ObliviousHarmony committed:
Updated the package-lock.json after fixing merge conflicts

ObliviousHarmony committed:
Corrected the tabs in the README.me

rrennick commented on tests/e2e/env/package.json:
```diff
@@ -41,6 +41,10 @@
   },
   "scripts": {
+    "clean": "rm -rf ./build ./build-module",
+    "compile": "node ./../bin/build.js",
```
Should this be moved to the main repo package.json. The build script won't exist in the `/node_modules/` folder when installed as a package.

ObliviousHarmony replied:
I feel like that's okay because we shouldn't be running any of these scripts from a parent project right? The build script is for preparing the packages for distribution so they aren't really relevant once distributed? The incentive for having it here is that we can run `lerna run build` and then each package in the monorepo is able to define its own build script as opposed to having a single universal script.

claudiosanches commented on tests/e2e/docker/initialize.sh:
```diff
@@ -5,3 +5,6 @@ echo "Initializing WooCommerce E2E"
 wp theme install twentynineteen --activate
 wp user create customer customer@woocommercecoree2etestsuite.com --user_pass=password --role=customer --path=/var/www/html
+
+# we cannot create API keys for the API, so we using basic auth, this plugin allows that.
```
In fact we can create API keys for our own REST API without having to install this plugin, unless you are planning to do some requests to WP endpoints instead of WooCommerce core endpoints.

ObliviousHarmony replied:
Initially I had planned on using the API authentication and even [wrote an interceptor for it](https://github.com/woocommerce/woocommerce/blob/e4d90f10c3643dc56d003384b407979ad17c21d4/tests/e2e/factories/src/framework/api/axios/axios-oauth-interceptor.ts)! The idea was that we would store a `consumer_key` and `consumer_secret` in the test suite config file and then generate that key when initializing the test suite. 
 Creating a new CLI command felt a little outside the scope of this PR though and so the basic auth plugin provided by WordPress seemed like a reasonable way to quickly get this working. 

I'll note that I did create an API key on my local site though and thoroughly test the interceptor so if we decide to go that route it will be a very easy change to make :smile:

rrennick commented:
@ObliviousHarmony Testing instructions should include any build instructions for the factories.

ObliviousHarmony commented:
@rrennick I've updated the instructions to be more helpful!

ObliviousHarmony committed:
Adjusted the publish configuration to respect the @woocommerce scope

juliaamosova commented:
@ObliviousHarmony this PR looks incredible. Great work! I love how it is now possible to create a product with just a few lines of code and get its ID. 

I've tried to run tests locally and came across the following error pretty much on every test that runs:

```
Failed: APIResponse {
      "data": APIError {
        "code": "woocommerce_rest_cannot_create",
        "data": Object {
          "status": 401,
        },
        "message": "Sorry, you are not allowed to create resources.",
      },
      "headers": Object {
        "access-control-allow-headers": "Authorization, Content-Type",
        "access-control-expose-headers": "X-WP-Total, X-WP-TotalPages",
        "connection": "close",
        "content-length": "123",
        "content-type": "application/json; charset=UTF-8",
        "date": "Fri, 17 Jul 2020 15:20:34 GMT",
        "link": "<http://localhost:8084/wp-json/>; rel=\"https://api.w.org/\"",
        "server": "Apache/2.4.38 (Debian)",
        "x-content-type-options": "nosniff",
        "x-powered-by": "PHP/7.3.17",
        "x-robots-tag": "noindex",
      },
      "status": 401,
    }
```

Can you please help me understand what could be going on? Is it something on my end? I followed the steps you outlined to test this PR. 

>Should I move all of the tests files into a tests folder? I'm a fan of having them next to the source files but it might be more consistent to do what we're doing elsewhere.

Currently, the tests are located in `woocommerce/tests/e2e/specs` and further organized in 3 folders:

- activate-and-setup
- front-end
- wp-admin

Do you propose that we eliminate the `e2e` folder and have tests in `woocommerce/tests/specs`? If so, I think that can be confusing to understand what these tests are at a glance. Plus `e2e` test folder contains config, env, and other folders - do you propose that we move them to `tests` as well? 

I am in favor of keeping the `e2e` folder and have code relevant to it there in order to keep things organized.

juliaamosova commented:
Also noting that OBW e2e test was failing (which is also a part of this PR). The fix for it is here: https://github.com/woocommerce/woocommerce/pull/27063. Will be a good idea to rebase as soon as https://github.com/woocommerce/woocommerce/pull/27063 is merged so that all e2e tests pass.

ObliviousHarmony commented:
> Can you please help me understand what could be going on?

I wonder if you have some leftover docker container and `npm run docker:up` isn't creating a new one with the correct initialization step? This requires the basic auth plugin to be installed and activated which might not be done. I would suggest perhaps destroying all of your existing docker containers if possible (or at least those related to the E2E suite) and trying it again!

> Do you propose that we eliminate the `e2e` folder and have tests in `woocommerce/tests/specs`? 

Sorry, I meant _just_ the tests for the factory package. In the `tests/e2e/factories/src/framework/api/` directory for example there is an `api-adapter.ts` and `api-adapter.spec.ts` file. I am partial to this kind of structure for tests because it makes finding the test file very easy but if desired I'm fine with moving them into a `tests/e2e/factories/tests` folder as well.

rrennick commented:
> This requires the basic auth plugin to be installed and activated which might not be done. I would suggest perhaps destroying all of your existing docker containers if possible

@ObliviousHarmony `docker:down` destroys the container. If the plugin needs to be installed then it should be added to the configuration to ensure it's installed.

rrennick reviewed (approved):
@ObliviousHarmony Nice work on this. I've been looking forward to seeing this. Locally I get the test failures due to the onboarding shipping changes in 4.3. Otherwise this works a treat.

@juliaamosova Did you do both `composer install` and `npm install` in the first step of the instructions?

senadir reviewed (approved):
Overall I think this is great, I a bit worried about the number of steps needed for someone to implement a basic factory like SimpleProduct, maybe those things can be abstracted more into simpler utilities?

ObliviousHarmony committed:
Update README.md

Changed `npm install` suggestion to better suggest saving the dependencies as devDependencies

ObliviousHarmony commented:
> Overall I think this is great, I a bit worried about the number of steps needed for someone to implement a basic factory like SimpleProduct, maybe those things can be abstracted more into simpler utilities?

I was thinking about this and I agree. While advanced consumers may want to have this level of control we should have a helper utility that completes all of the necessary setup and returns a ready-to-use registry.

juliaamosova commented:
Please don't wait for me to resolve the issues I had with running the tests. Don't want to be a blocker for the merge since the PR has 2 approvals already. I will try again in production. Likely it was something on my end that caused the errors.

ObliviousHarmony committed:
Merge branch 'master' into package/e2e-factories/product

ObliviousHarmony committed:
Pinned the dependencies that were not already pinned

ObliviousHarmony committed:
Fixed test that used PHPUnit function from newer version
